components/brave_wallet/browser/solana_message.cc

```diff
@@ -719,4 +720,97 @@ bool SolanaMessage::UsesDurableNonce() const {
 }
 
+bool SolanaMessage::UsesPriorityFee() const {
+  if (instructions_.empty()) {
+    return false;
+  }
+
+  auto instruction = instructions_[0];
+
+  // If the first instruction is advance nonce instruction,
+  // use the second instruction.
+  if (solana_ins_data_decoder::GetSystemInstructionType(
+          instruction.data(), instruction.GetProgramId()) ==
+      mojom::SolanaSystemInstruction::kAdvanceNonceAccount) {
+    if (instructions_.size() < 2) {
+      return false;
+    }
+
+    instruction = instructions_[1];
+  }
+
+  auto instruction_type =
+      solana_ins_data_decoder::GetComputeBudgetInstructionType(
+          instruction.data(), instruction.GetProgramId());
+  if (!instruction_type) {
+    return false;
+  }
+
+  return instruction_type ==
+         mojom::SolanaComputeBudgetInstruction::kSetComputeUnitLimit;
+}
+
+bool SolanaMessage::AddPriorityFee(uint32_t compute_units,
+                                   uint64_t fee_per_compute_unit) {
+  SolanaInstruction modify_compute_units_instruction =
+      solana::compute_budget_program::SetComputeUnitLimit(compute_units);
+  SolanaInstruction add_priority_fee_instruction =
+      solana::compute_budget_program::SetComputeUnitPrice(fee_per_compute_unit);
+
+  // Do not add a priority fee if there already is one added.
+  if (UsesPriorityFee()) {
+    return false;
+  }
+
+  if (UsesDurableNonce()) {
+    // The first instruction should remain the advance nonce instruction.
+    // https://solana.com/developers/guides/advanced/how-to-use-priority-fees#special-considerations
+    instructions_.insert(
+        instructions_.begin() + 1,
+        {modify_compute_units_instruction, add_priority_fee_instruction});
+  } else {
+    instructions_.insert(
+        instructions_.begin(),
+        {modify_compute_units_instruction, add_priority_fee_instruction});
+  }
+
+  // Rearrange the static account keys and generate a new SolanaMessageHeader
+  uint16_t num_required_signatures = 0;
+  uint16_t num_readonly_signed_accounts = 0;
+  uint16_t num_readonly_unsigned_accounts = 0;
+  std::vector<SolanaAccountMeta> unique_account_metas;
+  GetUniqueAccountMetas(fee_payer_, instructions_, &unique_account_metas);
+  std::vector<SolanaAddress> static_accounts;
+  for (const auto& meta : unique_account_metas) {
+    auto addr = SolanaAddress::FromBase58(meta.pubkey);
+    if (!addr) {
+      return false;
+    }
+
+    if (meta.is_signer) {
+      num_required_signatures++;
+    }
+    if (meta.is_signer && !meta.is_writable) {
+      num_readonly_signed_accounts++;
+    }
+    if (!meta.is_signer && !meta.is_writable) {
+      num_readonly_unsigned_accounts++;
+    }
+
+    if (num_required_signatures > UINT8_MAX ||
+        num_readonly_signed_accounts > UINT8_MAX ||
+        num_readonly_unsigned_accounts > UINT8_MAX ||
+        static_accounts.size() == UINT8_MAX) {
+      return false;
+    }
+    static_accounts.emplace_back(*addr);
+  }
+  static_account_keys_ = static_accounts;
+
```

Comment:
We should extract and share code in CreateLegacyMessage rather than copy the logic.